text/0036-ternary.md

```diff
@@ -0,0 +1,280 @@
+
+### `TritBuf`
+
+```rust
+struct TritBuf<T: RawEncodingBuf> { ... }
+```
+
+- Analogous to `String` or `Vec<T>`
+- Most common way to manipulate trits
+- Allows pushing, popping, and collecting trits
+- Implements `Deref<Target=Trits>` (in the same way that `Vec<T>` implements
+`Deref<[T]>`)
+
+### `RawEncoding`
+
+```rust
+trait RawEncoding { ... }
+```
+
+- Represents a raw trit buffer of some encoding
+- Common interface implemented by all trit encodings (`T1B1`, `T3B1`, `T5B1`,
+etc.)
+- Largely an implementation detail, not something you need to care about unless
+implementing your own encodings
+- Minimal implementation requirements: safety and most utility functionality is
+provided by `Trits` instead
+
+### `RawEncodingBuf`
+
+```rust
+trait RawEncodingBuf { ... }
+```
+
+- Buffer counterpart of `RawEncoding`, always associated with a specific
+`RawEncoding`
+- Distinct from `RawEncoding` to permit different buffer-like data structures
+in the future (linked lists, stack-allocated arrays, etc.)
+
+### `T1B1`/`T2B1`/`T3B1`/`T4B1`/`T5B1`
+
+```rust
+struct TXB1 { ... }
+```
+
+- Types that implement `RawEncoding`
+- Allow different encodings in `Trits` and `TritBuf` types.
+
+### `T1B1Buf`/`T2B1Buf`/`T3B1Buf`/`T4B1Buf`/`T5B1Buf`
+
+```rust
+struct TXB1Buf { ... }
+```
+
+- Types that implement `RawEncodingBuf`
+- Allow different encodings in `TritBuf` types.
+- Each type is associated with a `RawEncoding` type.
+
+## API
+
+The API makes up the body of this RFC. Due to its considerable length, this RFC
+simply refers to the documentation in question. You can find those docs in the
+`bee-p` (and later `bee`) repository.
+
+## Encodings
+
+`bee-ternary` supports many different trit encodings. Notable encodings are
+explained in the crate documentation.
+
+## Common Patterns
+
+When using the API, the most common types interacted with are `Trits` and
+`TritBuf`. These types are designed to play well with the rest of the Rust
+ecosystem. Here follows some examples of common patterns that you may wish to
+make use of.
+
+### Turning some `i8`s into a trit buffer
+
+```rust
+[-1, 0, 1, 1, -1, 0]
+	.iter()
+	.map(|x| Btrit::try_from(*x).unwrap())
+	.collect::<TritBuf>()
+```
+
+Alternatively, for the `T1B1` encoding only, you may directly reinterpret the
+`i8` slice.
+
+```rust
+let xs = [-1, 0, 1, 1, -1, 0];
+
+Trits::try_from_raw(&xs, xs.len())
+	.unwrap()
+```
+
+If you are *certain* that the slice contains *only* valid trits then it is
+possible to unsafely reinterpret with amortised O(1) cost (i.e: it's basically
+free). However, scenarios in which this is necessary and sound are exceedingly
+uncommon. If you find yourself doing this, ask yourself first whether it is
+necessary.
+
+```rust
+let xs = [-1, 0, 1, 1, -1, 0];
+
+unsafe { Trits::from_raw_unchecked(&xs, xs.len()) }
+```
+
+### Turning a trit slice into a tryte string
+
+```rust
+trits
+	.chunks(3)
+	.map(|trits| char::from(Tryte::from_trits([
+		trits.get(0),
```

Comment:
Does .get(x) still need the unwrap() or was the implementation changed?

Reply:
Yes, I'm just being very stupid today :smile: